polar-core/src/polar.rs

```diff
@@ -255,24 +255,26 @@ impl Polar {
                 .collect(),
         );
 
-        // Attach context to ResourceBlock errors.
+        // Attach context to ResourceBlock, SingletonVariable, and UnregisteredConstant errors.
         //
         // TODO(gj): can we attach context to *all* errors here since all errors will be parse-time
         // errors and so will have some source context to attach?
-        diagnostics = diagnostics
-            .into_iter()
-            .map(|d| {
-                if let Diagnostic::Error(ref e) = d {
-                    if let ErrorKind::Validation(
-                        ref e @ ValidationError::ResourceBlock { ref term, .. },
-                    ) = e.kind
-                    {
-                        return Diagnostic::Error(kb.set_error_context(term, e.clone()));
+        for diagnostic in &mut diagnostics {
+            if let Diagnostic::Error(e) = diagnostic {
+                use {ErrorKind::Validation, ValidationError::*};
+                match e.kind {
+                    Validation(ResourceBlock { ref term, .. })
+                    | Validation(SingletonVariable { ref term, .. })
+                    | Validation(UnregisteredConstant { ref term, .. }) => {
+                        *e = kb.set_error_context(term, e.clone());
```

Comment:
Conceptually we shouldn't need a clone here. Why doesn't `set_error_context` take a mut ref?

Reply:
dunno; preexisting API that I would like to change in a follow-up